lib/Extension/LanguageServerPhpCsFixer/Provider/PhpCsFixerDiagnosticsProvider.php

```diff
@@ -0,0 +1,274 @@
+                            $endPos = $lineLength
+                                ? new Position($lineNo - 1, $lineLength)
+                                : new Position($lineNo, 0);
+
+                            if ($replacedLines) {
+                                $firstLineA = $replacedLines[0]->getContent();
+                                $firstLineB = $changedLines[0]->getContent();
+                                $lastLineA = end($replacedLines)->getContent();
+                                $lastLineB = end($changedLines)->getContent();
+
+                                $startChars = StringSharedChars::startLength($firstLineA, $firstLineB);
+                                $endChars = StringSharedChars::endPos($lastLineA, $lastLineB);
+
+                                $startPos = new Position($startLineNo, $startChars);
+                                $endPos = new Position($lineNo - 1, $endChars);
+                            }
+
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range($startPos, $endPos),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $startLineNo = null;
+                            $changedLines = [];
+                        }
+
+                        if ($line->getType() === Line::UNCHANGED) {
+                            continue;
+                        }
+
+                        if ($line->getType() === Line::REMOVED) {
+                            $startLineNo = $lineNo;
+                            $changedLines[] = $line;
+
+                            continue;
+                        }
+
+                        $prevLine = $chunk->getLines()[$index - 1];
+
+                        if ($prevLine->getContent() === "\ No newline at end of file") {
+                            $contextLines = [];
+
+                            continue;
+                        }
+
+                        if ($line->getType() === Line::ADDED
+                            && $prevLine->getType() === Line::UNCHANGED
+                        ) {
+                            $diagnostics[] = Diagnostic::fromArray([
+                                'message' => yield $this->explainRule($rule),
+                                'range' => new Range(new Position($lineNo, 0), new Position($lineNo, 1)),
+                                'severity' => DiagnosticSeverity::WARNING,
+                                'source' => $this->name().':'.$rule
+                            ]);
+
+                            $contextLines = [];
+
+                            continue;
+                        }
+
+                        throw new LogicException('This is not intended to be reached');
```

Comment:
minor thing -- i would make the success action the _last thing_. The condition above can be `if ($line->getType() != Line::ADDED || $prevLine->getType() !== Line::UNCHANGED) { // throw exception }`. So all the conditions before it are "guards" including the "this should never happen"

Reply:
I'd do throw/exit early possible, but in this scenario I think it won't make sense.

The whole loop works by finding an appropriate case and continue on the loop and it's dependent on previous entries of the looped array.
Meaning that all successful scenarios are early the block, while potential incorrect behaviour is left as the last one - Making this change would mean we have few correct cases, an incorrect one followed by one more correct case

That's my reasoning behind it, but not too attached to it. If you prefer to reverse the if, then it's fine and I'll update that